aspects/schema.go

```diff
@@ -332,12 +357,21 @@ func (v *mapSchema) parseMapKeyType(raw json.RawMessage) (Schema, error) {
 		return schema, nil
 	}
 
-	// TODO: if type starts with $, check against user-defined types
-	if typ != "string" {
-		return nil, fmt.Errorf(`must be based on string but got %q`, typ)
+	if typ == "string" {
+		return &stringSchema{}, nil
+	}
+
+	if typ != "" && typ[0] == '$' {
+		if userType, ok := v.topSchema.userTypes[typ[1:]]; ok {
+			if _, ok := userType.(*stringSchema); !ok {
+				return nil, fmt.Errorf(`key type %q must be based on string`, typ[1:])
+			}
+			return userType, nil
+		}
+		return nil, fmt.Errorf(`cannot find user-defined type %q`, typ[1:])
```

Comment:
and part of this could share a helper?

Reply:
It could but it will add more code than remove because there's so little shared code:

```
diff --git a/aspects/schema.go b/aspects/schema.go
index 0d713d287a..dc60f4a4b1 100644
--- a/aspects/schema.go
+++ b/aspects/schema.go
@@ -148,15 +148,21 @@ func (s *StorageSchema) newTypeSchema(typ string) (parser, error) {
 		return &stringSchema{}, nil
 	default:
 		if typ != "" && typ[0] == '$' {
-			if userType, ok := s.userTypes[typ[1:]]; ok {
-				return userType, nil
-			}
-			return nil, fmt.Errorf("cannot find user-defined type %q", typ[1:])
+			return s.getUserType(typ[1:])
 		}
+
 		return nil, fmt.Errorf("cannot parse unknown type %q", typ)
 	}
 }
 
+func (s *StorageSchema) getUserType(ref string) (parser, error) {
+	if userType, ok := s.userTypes[ref]; ok {
+		return userType, nil
+	}
+
+	return nil, fmt.Errorf("cannot find user-defined type %q", ref)
+}
+
 type mapSchema struct {
 	// topSchema is the schema for the top-level schema which contains the user types.
 	topSchema *StorageSchema
@@ -362,13 +368,16 @@ func (v *mapSchema) parseMapKeyType(raw json.RawMessage) (Schema, error) {
 	}
 
 	if typ != "" && typ[0] == '$' {
-		if userType, ok := v.topSchema.userTypes[typ[1:]]; ok {
-			if _, ok := userType.(*stringSchema); !ok {
-				return nil, fmt.Errorf(`key type %q must be based on string`, typ[1:])
-			}
-			return userType, nil
+		userType, err := v.topSchema.getUserType(typ[1:])
+		if err != nil {
+			return nil, err
 		}
-		return nil, fmt.Errorf(`cannot find user-defined type %q`, typ[1:])
+
+		if _, ok := userType.(*stringSchema); !ok {
+			return nil, fmt.Errorf(`key type %q must be based on string`, typ[1:])
+		}
+		return userType, nil
+
 	}
 
 	return nil, fmt.Errorf(`keys must be based on string but got %q`, typ)
```

I can do it but just wanted to double check that you still prefer it